src/service/resources-impl.js

```diff
@@ -407,13 +410,29 @@ export class Resources {
    * @private
    */
   buildReadyResources_() {
+    // Avoid cases where elements add more elements inside of them
+    // and cause an infinite loop of building - see #3354 for details.
+    if (this.isCurrentlyBuildingPendingResources_) {
```

Comment:
This has a problem. If we shortcircuiting "may be we can build" signal - we need to reschedule it for later. E.g. consider this:

"A" is being build which adds "B". "B" is put in the queue and it's ready to be built, but its cycle ignored due to `this.isCurrentlyBuildingPendingResources_ == true`. Once "A" is finished building, "B" is still in the queue and now there's no signal to build it again, so it stays in the queue unbuilt.

One way we can deal with this is by using increment instead of boolean for `isCurrentlyBuildingPendingResources_ `. E.g.:

```
this.buildingPendingResourcesIndex_ = 0;

buildReadyResources_() {
  this.buildingPendingResourcesIndex_++;
  if (this.buildingPendingResourcesIndex_ > 1) {
   return;
  }
  try {
    this.buildReadyResourcesUnsafe_();
  } finally {
    this.buildingPendingResourcesIndex_--;
  }
  if (this.buildingPendingResourcesIndex_ > 0) {
    this.buildingPendingResourcesIndex_ = 0;
    this.buildReadyResources_();
  }
}
```

Makes sense?